cmd/installer/cli/upgrade.go

```
@@ -0,0 +1,439 @@
+	mustMarkFlagRequired(flagSet, "target")
+
+	flagSet.StringVar(&flags.airgapBundle, "airgap-bundle", "", "Path to the air gap bundle. If set, the upgrade will complete without internet access.")
+
+	flagSet.StringVar(&flags.overrides, "overrides", "", "File with an EmbeddedClusterConfig object to override the default configuration")
+	mustMarkFlagHidden(flagSet, "overrides")
+
+	flagSet.BoolVarP(&flags.assumeYes, "yes", "y", false, "Assume yes to all prompts.")
+	flagSet.SetNormalizeFunc(normalizeNoPromptToYes)
+
+	return flagSet
+}
+
+func addUpgradeAdminConsoleFlags(cmd *cobra.Command, flags *UpgradeCmdFlags) error {
+	cmd.Flags().StringVarP(&flags.licenseFile, "license", "l", "", "Path to the license file")
+	mustMarkFlagRequired(cmd.Flags(), "license")
+
+	return nil
+}
+
+func addUpgradeManagementConsoleFlags(cmd *cobra.Command, flags *UpgradeCmdFlags) error {
+	// default value of 0 indicates no user input - will use existing runtime config value
+	cmd.Flags().IntVar(&flags.managerPort, "manager-port", 0, "Port on which the Manager will be served")
+	return nil
+}
+
+func preRunUpgrade(ctx context.Context, flags UpgradeCmdFlags, upgradeConfig *upgradeConfig, rc runtimeconfig.RuntimeConfig, kcli client.Client) error {
+	// TODO: Add check for kubernetes target when kubernetes support is added
+	if !slices.Contains([]string{"linux"}, flags.target) {
+		return fmt.Errorf(`invalid --target (must be: "linux")`)
+	}
+
+	// Verify an installation exists and get the cluster ID
+	clusterID, err := getClusterID(ctx, kcli)
+	if err != nil {
+		return fmt.Errorf("could not get existing installation: %w", err)
+	}
+	upgradeConfig.clusterID = clusterID
+
+	// Verify that a data directory exists
+	dataDir := rc.EmbeddedClusterHomeDirectory()
+	if _, err := os.Stat(dataDir); os.IsNotExist(err) {
+		return fmt.Errorf("upgrade requires existing data directory from previous installation")
+	}
```

perhaps it is better not to mask the original error here. it may be help in diagnosing the issue.

```suggestion
	if _, err := os.Stat(dataDir); err != nil {
		return fmt.Errorf("failed to stat data directory: %w", err)
	}
```